jormungandr/src/utils/task.rs

```diff
@@ -100,25 +100,29 @@ impl Services {
         F: Send + 'static,
     {
         let now = Instant::now();
+        let logger = self
+            .logger
+            .new(o!(crate::log::KEY_TASK => name))
+            .into_erased();
         let thread_service_info = ThreadServiceInfo {
             name: name,
             up_time: now,
-            logger: self
-                .logger
-                .new(o!(crate::log::KEY_TASK => name))
-                .into_erased(),
+            logger: logger.clone(),
         };
-
-        let handler = thread::Builder::new()
+        let finish_notifier = self.finish_listener.notifier();
+        thread::Builder::new()
             .name(name.to_owned())
             // .stack_size(2 * 1024 * 1024)
             .spawn(move || {
-                info!(thread_service_info.logger, "starting task");
-                f(thread_service_info)
+                info!(logger, "starting task");
+                if let Err(error) = catch_unwind(AssertUnwindSafe(|| f(thread_service_info))) {
```

Comment:
That's another `AssertUnwindSafe` I'm not sure at all is safe. Have you tried to set the bound on `F` so that `catch_unwind` does not require this wrapper? I tried to do it previously, but there was a lot to push through that closure boundary that did not confrom, so I gave up.

Reply:
@mzabaluev So I did a little research.

I've tried forcing tasks to be `UnwindSafe` and it's not feasible. Huge part of `futures` and `tokio` infrastructure is `!UnwindSafe`. E.g. `tokio::Lock` and `futures::mpsc`, which aren't realistically replaceable. `async_std` provides `UnwindSafe` `Mutex`, which may be an alternative to `tokio::Lock`, but its implementation of `channel` is still `!UnwindSafe`.

But this doesn't seem to be a problem at all in context of thread executor. The unwind boundary makes sense only in context of current thread. If thread is created and its content panics, its closure doesn't have to be unwind safe, thread boundary is much stronger and can replace unwind boundary. E.g. `std::thread::spawn` doesn't need closure to be `UnwindSafe`, the fact that it's sent to another thread is enough even though thread technically catches panics. Here we can guarantee that nothing not unwind safe is used after panic is caught, so it's perfectly safe.

It's the same in case of tokio tasks. Whenever there's a panic, we catch it and finish execution, but because it's the only future on executor, it stops right away guaranteeing to never touch `!UnwindSafe` data anymore.

Reply:
We should not need to run `thread`s in the module. The last one still using a non `tokio` task is `client.rs` I believe. Better do the little leap and finish the client.rs work to make it asynchronous. That way we can remove this function and all the unsafe. Also this would simplify this module and the future work on porting to standard future APIs.

Reply:
What would be the plan to catch panics once everything is running in an executor's thread pool?
We want to exit as soon as it happens, but the behavior without `.catch_unwind` is that the pool recovers from loss of a thread dropping the parked task, or is it not?

Reply:
Or did you mean we do run tasks as futures wrapped in the `.catch_unwind()` combinator, not in separate threads?

Reply:
Yes, last one. It's time to clean up the last module and to make it async too, in a future task.

Reply:
Tokio runtime always catches panics and there's nothing we can do about it. What we can do (and I propose it here) is shut down everything whenever any future stops for whatever reason. `.catch_unwind` is used only to log the panic for the user.

Right now we have 2 tasks running on thread executor: client and REST.